src/pymor/algorithms/hapod.py

```diff
@@ -6,66 +6,122 @@
 
     @property
     def depth(self):
-        def get_depth(node):
-            children = self.children(node)
-            if children:
-                return 1 + max(get_depth(c) for c in children)
-            else:
-                return 1
-        return get_depth(self.root)
-
-    def is_leaf(self, node):
-        return not self.children(node)
-
+        max_level = 0
+        for _, level in self.traverse(True):
+            max_level = max(max_level, level)
+        return max_level + 1
 
-class IncHAPODTree(Tree):
-
-    def __init__(self, steps):
-        self.steps = steps
-        self.root = steps
+    @property
+    def is_leaf(self):
+        return not self.children
 
-    def children(self, node):
-        if node < 0:
-            return ()
-        elif node == 1:
-            return (-1,)
+    @property
+    def is_root(self):
+        return not self.parent
+
+    def traverse(self, return_level=False):
+        current_node = self
+        last_child = None
+        level = 0
+        while True:
+            if last_child is None:
+                if return_level:
+                    yield current_node, level
+                else:
+                    yield current_node
+            if current_node.children:
+                if last_child is None:
+                    current_node = current_node.children[0]
+                    level += 1
+                    continue
+                else:
+                    last_child_pos = current_node.children.index(last_child)
+                    if last_child_pos + 1 < len(current_node.children):
+                        current_node = current_node.children[last_child_pos+1]
+                        last_child = None
+                        level += 1
+                        continue
+            if not current_node.parent:
+                return
+            last_child = current_node
+            current_node = current_node.parent
+            level -= 1
+
+    def __str__(self):
+        lines = []
+        for node, level in self.traverse(True):
+            line = ''
+            if node.parent:
+                p = node.parent
+                while p.parent:
+                    if p.parent.children.index(p) + 1 == len(p.parent.children):
+                        line = '  ' + line
+                    else:
+                        line = '| ' + line
+                    p = p.parent
+                line += '+-o'
+            else:
+                line += 'o'
+            if node.tag is not None:
+                line += f' {node.tag}'
+            if node.after:
+                line += f' (after {",".join(str(a) for a in node.after)})'
+            lines.append(line)
+        return '\n'.join(lines)
+
+
+def inc_hapod_tree(steps):
+    tree = node = Node()
+    for step in range(steps)[::-1]:
+        # add leaf node for a new snapshot and set local_eps to 0 to prevent computing a POD
+        node.add_child(step, after=(step-1,) if step > 0 else None)
```

Comment:
specifying `tag` here helps me reading it better
```suggestion
        node.add_child(tag=step, after=(step-1,) if step > 0 else None)
```